qiskit_alice_bob_provider/plugins/state_preparation.py

```diff
@@ -313,41 +337,91 @@ def pass_manager(
+        else:
+            # In this case, our basis target consists of Clifford + T gate,
+            # i.e. {cx, h, s, t}. It does not contain unitary rotations
+            # (e.g. rz).
+            #
+            # This is a universal set of gates, but it requires synthesis
+            # (e.g. with SK algorithm), for transpilation to succeed. This
+            # synthesis (that transforms rotations into discrete gates of our
+            # basis target) is done during the CustomUnitarySynthesis pass.
+            #
+            # Unfortunately, qiskit BasisTranslator pass does not handle
+            # synthesis. It tries to match existing gates in the circuit to
+            # gates in the target basis, but only for exact equivalence (doing
+            # a graph search).
+            #
+            # Therefore, as a workaround, when the target basis does not
+            # support unitary rotations natively, we
+            #   - add the 'rz' to the BasisTranslator target basis set, to
+            #     trick the algorithm into thinking that rotations are
+            #     supported by the target.
+            #   - add a 2nd CustomUnitarySynthesis pass after that, to get
+            #     rid of any 'rz' pass generated in the pass above.
+            #
+            # Example :
+            # > Input circuit ('cs')
+            # q_0: ──■──
+            #      ┌─┴─┐
+            # q_1: ┤ S ├
+            #      └───┘
+            #
+            # > output of BasisTranslator, with 'rz' in target :
+            #      ┌─────────┐
+            # q_0: ┤ Rz(π/4) ├──■────────────────■─────────────
+            #      └─────────┘┌─┴─┐┌──────────┐┌─┴─┐┌─────────┐
+            # q_1: ───────────┤ X ├┤ Rz(-π/4) ├┤ X ├┤ Rz(π/4) ├
+            #                 └───┘└──────────┘└───┘└─────────┘
+            #
+            # > output of BasisTranslator, without 'rz' in target :
+            # TranspilerError: "Unable to translate the operations..."
+            #
+            # > output of the 2nd CustomUnitarySynthesis :
+            #      ┌───┐
+            # q_0: ┤ T ├──■───────────■───────
+            #      └───┘┌─┴─┐┌─────┐┌─┴─┐┌───┐
+            # q_1: ─────┤ X ├┤ Tdg ├┤ X ├┤ T ├
+            #           └───┘└─────┘└───┘└───┘
+            need_synthesis = True
+
+            # Ideally, we would use the `target_basis` argument of
+            # BasisTranslator to specify we want to add the "rz" gate.
+            # Unfortunately, in the .run method, this argument is ignored in
+            # favor of the target.target_basis when the target is defined.
+            # Therefore, we create a copy of the current target and modify
+            # its set of supported instructions just for this pass.
+            basis_translator_target = copy(pass_manager_config.target)
+            # pylint: disable=protected-access
+            basis_translator_target._gate_map = copy(
+                pass_manager_config.target._gate_map
+            )
```

Comment:
Could it be simpler and safer to just do a `deepcopy()` of `pass_manager_config.target` here ?

Reply:
Good question !

I was afraid of side effects `deepcopy` could have on other attributes of the target, in particular on objects with circular references (e.g. `target._coupling_graph`), thinking it could break some links.

But [looking at python docs](https://docs.python.org/3/library/copy.html), it's true that it should be handled by the `memo` dict.

Unit tests are still happy if I use `deepcopy`, so it could be an option. Do you think it would be safer ?

Reply:
The behavior I was thinking of is that with `copy()` if you change a nested value in `pass_manager_config` it will probably be changed also in `basis_translator_target`as it's a new object made of references. But finally maybe it's not an issue and it will even keep the initial behavior of the passed `pass_manager_config` object ?